### PR TITLE
Helper functions for list item fields

### PR DESCRIPTION
**Pull request by ebellbog**
Added two fully implemented and tested functions for dealing with SalesforceIQ's somewhat convoluted list item field syntax: formatListFields and parseListFields. These each make an API call to SalesforceIQ to obtain the additional data needed to "encode" and "decode" between simple key/value stores and SFIQ syntax, which relies on unnamed numerical indices. Also added stub for an updateListFields function, which would combine functionality of both of these helpers and save the system from making redundant API calls. This still needs implementation.

### CODE REVIEW

**Comment by ebellbog** [comment]
@wchest & @matthewheck: I would recommend reviewing and merging this PR first, since my other PRs rely on it.

File: lib/salesforceiq.js [hunk 1]
```diff
@@ -90,6 +90,7 @@ var SalesforceIQ = (function() {
     });
   };
 
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove extra space.

File: lib/salesforceiq.js [hunk 2]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+   * (c.f. https://api.salesforceiq.com/#/curl#documentation_lists_list-items_list-item-object)
+   *
+   * @param {string}                  listId  Reference id for the list containing these fields
+   * @param {Object.<string, string>} data    Simple key/value store of field data
+   * @param {Function}                cb      Callback returning error info and formatted fields
+   */
+  SalesforceIQ.prototype.formatListFields = function(listId, data, cb) {
+    this.getList(listId, function(err, list) {
+      if (err) {
+        // REF console.log(err);
```
Comment (matthewheck) [inline lib/salesforceiq.js]: let's remove this.

File: lib/salesforceiq.js [hunk 3]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+          _.forEach(field.listOptions, function(option) {
+            optionIds[option.display] = option.id;
+          });
+
+          if (optionIds) {
+            fieldOptionIds[field.name] = optionIds;
+          }
+        });
+
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove extra space

File: lib/salesforceiq.js [hunk 4]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+                return;
+              }
+            }
+            fieldValues[nameId] = [{raw: value}];
+          } else {
+            invalidFieldNames.push(name);
+          }
+        });
+
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove extra space.

File: lib/salesforceiq.js [hunk 5]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+        if (returnError) {
+          cb(err, data);
+        } else {
+          cb(null, data);
+        }
+      }
+    });
+  };
+
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove extra line

File: lib/salesforceiq.js [hunk 6]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+      if (err) {
+        // REF console.log(err);
+        cb(err);
+      } else {
+        // @TODO: implement functionality
+      }
+    });
+  };
+
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove  extra line

File: lib/salesforceiq.js [hunk 7]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+        if (returnError) {
+          cb(err, fieldValues);
+        } else {
+          cb(null, fieldValues);
+        }
+      }
+    });
+  };
+
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove extra line

File: lib/salesforceiq.js [hunk 8]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+          _.forEach(field.listOptions, function(option) {
+            optionNames[option.id] = option.display;
+          });
+
+          if (optionNames) {
+            idToFieldOption[field.id] = optionNames;
+          }
+        });
+
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove extra line

File: lib/salesforceiq.js [hunk 9]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+                return;
+              }
+            }
+            data[name] = valueOrValueId;
+          } else {
+            invalidFieldNameIds.push(id);
+          }
+        });
+
+
```
Comment (matthewheck) [inline lib/salesforceiq.js]: remove extra line

File: lib/salesforceiq.js [hunk 10]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
     makeRequest('lists/' + listId, req, cb);
   };
 
+  /**
+   * Format fields for list item to match SalesforceIQ API specs,
+   * using ids pulled from the associated list
+   * (c.f. https://api.salesforceiq.com/#/curl#documentation_lists_list-items_list-item-object)
+   *
+   * @param {string}                  listId  Reference id for the list containing these fields
+   * @param {Object.<string, string>} data    Simple key/value store of field data
```
Comment (matthewheck) [inline lib/salesforceiq.js:455]: I'm not familiar with documenting param types like this. Are you trying to do something more like this?
http://usejsdoc.org/tags-param.html#parameters-with-properties

File: lib/salesforceiq.js [hunk 11]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+        var invalidFieldOptions = {};
+
+        _.forEach(data, function(value, name) {
+          var nameId = fieldNameIds[name];
+          if (nameId) {
+            var pickList = fieldOptionIds[name];
+            if (!_.isEmpty(pickList)) {
+              var valueId = pickList[value];
+              if (valueId) {
+                value=valueId;
```
Comment (matthewheck) [inline lib/salesforceiq.js]: let's put a space before and after =. It's definitely aesthetic, but is going to be part of our standards in ESLint, so might as well get started!

File: lib/salesforceiq.js [hunk 12]
```diff
@@ -447,6 +448,196 @@ var SalesforceIQ = (function() {
...
+   *
+   * @param {string}                  listId  Reference id for the list containing these fields
+   * @param {Object.<string, string>} data    Simple key/value store of field data
+   * @param {Function}                cb      Callback returning error info and formatted fields
+   */
+  SalesforceIQ.prototype.formatListFields = function(listId, data, cb) {
+    this.getList(listId, function(err, list) {
+      if (err) {
+        // REF console.log(err);
+        cb(err);
```
Comment (matthewheck) [inline lib/salesforceiq.js]: This was flagged as an error with ESLint and I learned something new by reading about it!

http://eslint.org/docs/rules/callback-return

**Comment by matthewheck** [comment]
Almost all of my changes are cosmetic, but I took a look at the larger file and there are definitely some practices in there that aren't quite right. After we get your stuff in, we should do another round of review just to clean up the code and follow best practices better. Also, another step will be increasing code coverage (particularly by testing your code).

Great job getting this all together! Pretty much good to go on this!

File: package.json [hunk 13]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "salesforceiq",
-  "version": "1.0.0",
-  "description": "SalesforceIQ API wrapper for NodeJS",
+  "name": "tc-sfiq",
+  "version": "1.1.0",
+  "description": "Fork of SalesforceIQ API wrapper for NodeJS",
```
Comment (matthewheck) [inline package.json]: Let's remove Fork of on this

File: package.json [hunk 14]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "salesforceiq",
-  "version": "1.0.0",
-  "description": "SalesforceIQ API wrapper for NodeJS",
+  "name": "tc-sfiq",
```
Comment (matthewheck) [inline package.json]: and change this back to salesforceiq

File: lib/salesforceiq.js (hunk 10 above)
Comment (matthewheck) [inline lib/salesforceiq.js:455]: Any update on what's going on here?

**Comment by matthewheck** [comment]
Looking good, just three more very small things

File: lib/salesforceiq.js (hunk 10 above)
Comment (ebellbog) [inline lib/salesforceiq.js:455]: Yes! Sorry, found the documentation. Here it is: http://usejsdoc.org/tags-type.html

Quoted:
"An object with string keys and number values: {Object.<string, number>}"

I wanted to indicate that the object was a dictionary taking string keys and returning string values.

**Pull request merged**